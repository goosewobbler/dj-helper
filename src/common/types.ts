import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';

enum ModuleType {
  View,
  ViewCSS,
  Data,
}

enum ComponentType {
  Page = 1,
  View,
  Data,
}

enum ComponentState {
  Stopped = 1,
  Starting,
  Installing,
  Building,
  Running,
  Linking,
}

interface AppState {
  components: ComponentData[];
  ui: {
    componentToClone?: string;
    editors: string[];
    selectedComponent?: string;
    filter?: string;
    updating?: boolean;
    updated?: boolean;
    showDialog?: string;
    hideDialog?: boolean;
  };
}

interface Component {
  bump(type: BumpType): Promise<void>;
  fetchDetails(): Promise<void>;
  getName(): string;
  getDirectoryName(): string;
  getDisplayName(): string;
  getType(): ComponentType;
  getURL(): string | null;
  getFavourite(): boolean;
  getHistory(): string[];
  getUseCache(): boolean;
  setFavourite(favourite: boolean): Promise<void>;
  setUseCache(useCache: boolean): Promise<void>;
  getDependencies(): ComponentDependency[];
  getDependenciesSummary(): Promise<{ name: string }[]>;
  getLatestVersion(): Promise<string>;
  getLinking(): string[];
  getVersions(): Versions;
  getState(): ComponentState;
  getPromoting(): string | null;
  getPromotionFailure(): string | null;
  getRendererType(): string;
  promote(environment: string): Promise<void>;
  openInEditor(): Promise<void>;
  reinstall(): Promise<void>;
  link(dependency: string): Promise<void>;
  unlink(dependency: string): Promise<void>;
  makeLinkable(): Promise<void>;
  build(isSassOnly?: boolean, path?: string): Promise<void>;
  setPagePort(pagePort: number): void;
  start(): Promise<void>;
  stop(): Promise<void>;
  request(props: LooseObject, history: boolean): Promise<Response>;
}

interface ComponentsData {
  components: ComponentData[];
  editors: string[];
}

enum BumpType {
  patch = 'patch',
  minor = 'minor',
}

type ComponentMatch = string | { matched: string };

interface ComponentData {
  name: string;
  displayName: string;
  highlighted?: React.ReactElement[];
  state: ComponentState;
  favourite: boolean;
  history?: string[];
  url?: string;
  type?: ComponentType;
  dependencies?: ComponentDependency[];
  linking?: string[];
  promoting?: string;
  promotionFailure?: string;
  useCache: boolean;
  versions?: Versions;
  rendererType: string;
  alternatives?: string[];
  matches?: ComponentMatch[];
}

interface ComponentDependency {
  name: string;
  displayName: string;
  has: string | null;
  latest: string | null;
  linked: boolean;
  outdated: boolean;
  version: string | null;
  rendererType: string | null;
}

type GraphNode = {
  id: number;
  name: string;
};

type GraphData = {
  edges: {
    from: number;
    to: number;
  }[];
  nodes: GraphNode[];
};

interface Response {
  body: string;
  headers: Headers;
  statusCode: number;
}

type StringObject = { [Key: string]: string };

interface Shrinkwrap {
  dependencies: {
    [Key: string]: StringObject;
  };
}

interface Package {
  name: string;
  description: string;
  homepage: string;
  scripts: StringObject;
  dependencies: StringObject;
  devDependencies: StringObject;
  version: string;
  morph: { rendererType: string };
}

type AnyObject = Record<string, unknown>;

interface LooseObject {
  [Key: string]: AnyObject | [] | string | number | boolean;
}

interface AppStatus {
  currentVersion: string;
  updateAvailable: string;
  updated: boolean;
  updating: boolean;
}

interface FileSystem {
  exists(path: string): Promise<boolean>;
  getPackageDirectories(directory: string): string[];
  readFile(path: string): Promise<string>;
  writeFile(path: string, contents: string): Promise<void>;
  symbolicLinkExists(path: string): Promise<boolean>;
  copyDirectory(from: string, to: string, filter: boolean): Promise<void>;
  deleteDirectory(directory: string): void;
  watchDirectory(directory: string, callback: (path: string) => void): Promise<void>;
  moveDirectory(from: string, to: string): void;
  createSymlink(from: string, to: string): void;
  removeSymlink(path: string): void;
}

interface GitSystem {
  checkoutMaster(directory: string): Promise<void>;
  checkoutExistingBranch(directory: string, branchName: string): Promise<void>;
  checkoutNewBranch(directory: string, branchName: string): Promise<void>;
  commit(directory: string, message: string): Promise<void>;
  getCurrentBranch(directory: string): Promise<string>;
  getRandomBranchName(): Promise<string>;
  push(directory: string, branchName: string): Promise<void>;
  readyToCommit(directory: string): Promise<boolean>;
  stageFile(directory: string, path: string): Promise<void>;
}

interface MorphSystem {
  getShrinkwrap(name: string): Promise<{ [Key: string]: string }>;
  getVersionOnEnvironment(name: string, environment: string): Promise<string>;
  promote(name: string, environment: string): Promise<void>;
}

interface NetworkSystem {
  get(url: string): Promise<Response | undefined>;
}

interface ProcessSystem {
  getCurrentWorkingDirectory(): Promise<string>;
  log(message: string): void;
  open(url: string): void;
  runToCompletion(
    directory: string,
    command: string,
    onOutput: (message: string) => void,
    onError: (message: string) => void,
  ): Promise<void>;
  runUntilStopped(
    directory: string,
    command: string,
    onOutput: (message: string) => void,
    onError: (message: string) => void,
  ): Promise<() => Promise<void>>;
}

interface System {
  file: FileSystem;
  git: GitSystem;
  morph: MorphSystem;
  network: NetworkSystem;
  process: ProcessSystem;
}

type Dispatch = ThunkDispatch<AppState, undefined, AnyAction>;

type StoredValue = string[] | string | number | boolean;

interface ValueStore {
  [Key: string]: StoredValue;
}

interface Store {
  get(key: string): StoredValue | null;
  set(key: string, value: StoredValue | null): Promise<void>;
}

// type EnvironmentValues = 'int' | 'local' | 'live' | 'test';
// enum EnvironmentValues {
//   int = 'int',
//   local = 'local',
//   live = 'live',
//   test = 'test',
// }

type Versions = {
  [key: string]: string | null;
};
// type Versions = { [key in EnvironmentValues]: string | null };

interface Service {
  bump(name: string, type: BumpType): Promise<void>;
  build(name: string): Promise<void>;
  clone(name: string, cloneName: string, options: { description: string }): Promise<void>;
  create(name: string, type: ModuleType, options: { description: string }): Promise<void>;
  fetchDetails(name: string): Promise<void>;
  getComponentsData(): ComponentsData;
  getComponentsSummaryData(): ComponentsData;
  getDependantGraph(name: string): GraphData;
  getDependencyGraph(name: string): GraphData;
  link(name: string, dependency: string): Promise<void>;
  openInEditor(name: string): Promise<void>;
  promote(name: string, environment: string): Promise<void>;
  reinstall(name: string): Promise<void>;
  request(name: string, props: LooseObject, history: boolean): Promise<Response>;
  setFavourite(name: string, favourite: boolean): Promise<void>;
  setUseCache(name: string, useCache: boolean): Promise<void>;
  start(name: string): Promise<void>;
  stop(name: string): Promise<void>;
  unlink(name: string, dependency: string): Promise<void>;
}

export type ComponentHandlers = {
  buildComponent(name: string): void;
  bumpComponent(name: string, type: string): void;
  showCloneComponentDialog(sourceComponent: string): void;
  installComponent(name: string): void;
  linkComponent(name: string, dependency: string): void;
  promoteComponent(name: string, dependency: string): void;
  updateAndSelectComponent(name: string): void;
  setUseCacheOnComponent(name: string, value: boolean): void;
  unlinkComponent(name: string, dependency: string): void;
  openInCode(name: string): void;
};

export {
  AnyObject,
  Shrinkwrap,
  StringObject,
  Versions,
  LooseObject,
  BumpType,
  Component,
  ComponentData,
  ComponentDependency,
  ComponentMatch,
  ComponentState,
  ComponentType,
  ComponentsData,
  GraphData,
  GraphNode,
  ModuleType,
  Response,
  Package,
  Service,
  AppStatus,
  AppState,
  Dispatch,
  System,
  FileSystem,
  GitSystem,
  MorphSystem,
  NetworkSystem,
  ProcessSystem,
  Store,
  StoredValue,
  ValueStore,
};
