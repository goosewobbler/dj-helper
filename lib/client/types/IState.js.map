{"version":3,"file":"IState.js","sourceRoot":"","sources":["../../../src/client/types/IState.ts"],"names":[],"mappings":""}