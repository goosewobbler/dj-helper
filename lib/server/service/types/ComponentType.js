"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ComponentType;
(function (ComponentType) {
    ComponentType[ComponentType["Page"] = 1] = "Page";
    ComponentType[ComponentType["View"] = 2] = "View";
    ComponentType[ComponentType["Data"] = 3] = "Data";
})(ComponentType || (ComponentType = {}));
exports.default = ComponentType;
//# sourceMappingURL=ComponentType.js.map