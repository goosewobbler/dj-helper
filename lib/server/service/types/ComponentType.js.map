{"version":3,"file":"ComponentType.js","sourceRoot":"","sources":["../../../../src/server/service/types/ComponentType.ts"],"names":[],"mappings":";;AAAA,IAAK,aAIJ;AAJD,WAAK,aAAa;IAChB,iDAAQ,CAAA;IACR,iDAAI,CAAA;IACJ,iDAAI,CAAA;AACN,CAAC,EAJI,aAAa,KAAb,aAAa,QAIjB;AAED,kBAAe,aAAa,CAAC"}