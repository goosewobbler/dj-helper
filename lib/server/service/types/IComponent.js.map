{"version":3,"file":"IComponent.js","sourceRoot":"","sources":["../../../../src/server/service/types/IComponent.ts"],"names":[],"mappings":""}