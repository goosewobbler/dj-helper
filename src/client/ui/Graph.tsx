import glamorous from 'glamorous';
import * as React from 'react';

import IGraphData from '../../server/types/IGraphData';
import ITheme from '../../types/ITheme';
import LoadingIcon from './icon/LoadingIcon';

const GraphVis = require('react-graph-vis').default; // tslint:disable-line no-var-requires

const ContaingingDiv = glamorous.div({
  bottom: 0,
  left: 0,
  position: 'absolute',
  right: 0,
  top: 0,
});

const LoadingDiv = glamorous.div({
  alignItems: 'center',
  display: 'flex',
  height: '100%',
  justifyContent: 'center',
  width: '100%',
});

const LoadingIconContainer = glamorous.div({
  height: '32px',
  width: '32px',
});

interface IGraphProps {
  theme: ITheme;
  url: string;
  down: boolean;
  onSelect(name: string): any;
}

interface IGraphState {
  data: IGraphData;
}

const convertData = (data: IGraphData) => {
  const newNodes = data.nodes.map(node => ({ id: node.id, label: node.name.replace(/^bbc-morph-/, '') }));
  return {
    edges: data.edges,
    nodes: newNodes,
  };
};

class Graph extends React.PureComponent<IGraphProps, IGraphState> {
  constructor(props: IGraphProps) {
    super(props);

    this.state = {
      data: null,
    };
  }

  public componentDidMount() {
    this.update();
  }

  public componentDidUpdate() {
    if (!this.state.data) {
      this.update();
    }
  }

  public componentWillReceiveProps(nextProps: IGraphProps) {
    if (nextProps.url !== this.props.url) {
      this.setState({ data: null });
    }
  }

  public render() {
    return <ContaingingDiv>{this.renderGraph()}</ContaingingDiv>;
  }

  private update() {
    fetch(this.props.url)
      .then(res => res.json())
      .then(data => {
        this.setState({ data });
      })
      .catch(console.error);
  }

  private renderGraph() {
    if (this.state.data) {
      const options = {
        edges: {
          arrows: {
            from: {
              enabled: !this.props.down,
              scaleFactor: 0.5,
            },
            to: {
              enabled: this.props.down,
              scaleFactor: 0.5,
            },
          },
          smooth: {
            forceDirection: 'vertical',
            roundness: 1,
            type: 'cubicBezier',
          },
        },
        layout: {
          hierarchical: {
            direction: this.props.down ? 'UD' : 'DU',
            edgeMinimization: true,
            nodeSpacing: 350,
            parentCentralization: true,
            sortMethod: 'directed',
          },
          improvedLayout: true,
        },
        nodes: {
          color: {
            background: this.props.theme.tertiaryBackgroundColour,
            border: this.props.theme.primaryTextColour,
            highlight: {
              background: this.props.theme.tertiaryBackgroundColour,
              border: this.props.theme.highlightColour,
            },
          },
          font: {
            color: this.props.theme.tertiaryTextColour,
            face: this.props.theme.font,
          },
          scaling: {
            max: 30,
            min: 30,
          },
          shape: 'box',
          size: 30,
        },
        physics: {
          enabled: false,
        },
      };

      const events = {
        doubleClick: (event: any) => {
          const id = event.nodes[0];
          const name = this.state.data.nodes.find(n => n.id === id).name;
          this.props.onSelect(name);
        },
      };

      const currentNode = this.state.data.nodes[0].id;

      return (
        <GraphVis
          ref={(g: any) => {
            if (g) {
              g.Network.focus(currentNode, {
                scale: 1,
              });
            }
          }}
          graph={convertData(this.state.data)}
          events={events}
          options={options}
        />
      );
    }

    return (
      <LoadingDiv>
        <LoadingIconContainer>
          <LoadingIcon colour={this.props.theme.primaryTextColour} />
        </LoadingIconContainer>
      </LoadingDiv>
    );
  }
}

export default Graph;
