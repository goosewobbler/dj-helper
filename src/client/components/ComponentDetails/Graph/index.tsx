import * as React from 'react';
import GraphVis from 'react-graph-vis';

import { GraphData } from '../../../../common/types';
import LoadingIcon from '../../LoadingIcon';
import { logError } from '../../../../server/helpers/console';

interface GraphProps {
  url: string;
  down: boolean;
  onSelect(name: string): any;
}

interface GraphState {
  data: GraphData;
}

const convertData = (data: GraphData) => {
  const newNodes = data.nodes.map(node => ({ id: node.id, label: node.name.replace(/^bbc-morph-/, '') }));
  return {
    edges: data.edges,
    nodes: newNodes,
  };
};

class Graph extends React.PureComponent<GraphProps, GraphState> {
  constructor(props: GraphProps) {
    super(props);

    this.state = {
      data: null,
    };
  }

  public componentDidMount() {
    this.update();
  }

  public componentDidUpdate() {
    if (!this.state.data) {
      this.update();
    }
  }

  public componentWillReceiveProps(nextProps: GraphProps) {
    if (nextProps.url !== this.props.url) {
      this.setState({ data: null });
    }
  }

  public render() {
    return <div>{this.renderGraph()}</div>;
  }

  private update() {
    fetch(this.props.url)
      .then(res => res.json())
      .then(data => {
        this.setState({ data });
      })
      .catch(logError);
  }

  private renderGraph() {
    if (this.state.data) {
      const options = {
        edges: {
          arrows: {
            from: {
              enabled: !this.props.down,
              scaleFactor: 0.5,
            },
            to: {
              enabled: this.props.down,
              scaleFactor: 0.5,
            },
          },
          smooth: {
            forceDirection: 'vertical',
            roundness: 1,
            type: 'cubicBezier',
          },
        },
        layout: {
          hierarchical: {
            direction: this.props.down ? 'UD' : 'DU',
            edgeMinimization: true,
            nodeSpacing: 350,
            parentCentralization: true,
            sortMethod: 'directed',
          },
          improvedLayout: true,
        },
        nodes: {
          color: {},
          font: {},
          scaling: {
            max: 30,
            min: 30,
          },
          shape: 'box',
          size: 30,
        },
        physics: {
          enabled: false,
        },
      };

      const events = {
        doubleClick: (event: any) => {
          const id = event.nodes[0];
          const { name } = this.state.data.nodes.find(n => n.id === id);
          this.props.onSelect(name);
        },
      };

      const currentNode = this.state.data.nodes[0].id;

      return (
        <GraphVis
          ref={(g: any) => {
            if (g) {
              g.Network.focus(currentNode, {
                scale: 1,
              });
            }
          }}
          graph={convertData(this.state.data)}
          events={events}
          options={options}
        />
      );
    }

    return (
      <div className="loading">
        <div>
          <LoadingIcon />
        </div>
      </div>
    );
  }
}

export default Graph;
