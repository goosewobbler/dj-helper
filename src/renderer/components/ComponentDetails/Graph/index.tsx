import { ipcRenderer } from 'electron';
import * as React from 'react';
import GraphVis from 'react-graph-vis';

import { GraphData, GraphNode, AnyObject } from '../../../../common/types';
import { LoadingIcon } from '../../LoadingIcon';

type GraphProps = {
  type: 'dependency' | 'dependant';
  componentName: string;
  onSelect(name: string): void;
};

const convertData = (data: GraphData): AnyObject => {
  const newNodes = data.nodes.map((node): AnyObject => ({ id: node.id, label: node.name.replace(/^bbc-morph-/, '') }));
  return {
    edges: data.edges,
    nodes: newNodes,
  };
};

const renderLoadingIcon = (): React.ReactNode => (
  <div className="flex items-center justify-center w-full h-full loading">
    <div className="w-8 h-8">
      <LoadingIcon />
    </div>
  </div>
);

const renderGraph = (data: GraphData, onSelect: GraphProps['onSelect'], type: GraphProps['type']): React.ReactNode => {
  const isDependencyGraph = type === 'dependency';
  const options = {
    edges: {
      arrows: {
        from: {
          enabled: !isDependencyGraph,
          scaleFactor: 0.5,
        },
        to: {
          enabled: isDependencyGraph,
          scaleFactor: 0.5,
        },
      },
      smooth: {
        forceDirection: 'vertical',
        roundness: 1,
        type: 'cubicBezier',
      },
    },
    layout: {
      hierarchical: {
        direction: isDependencyGraph ? 'UD' : 'DU',
        edgeMinimization: true,
        nodeSpacing: 350,
        parentCentralization: true,
        sortMethod: 'directed',
      },
      improvedLayout: true,
    },
    nodes: {
      color: {},
      font: {},
      scaling: {
        max: 30,
        min: 30,
      },
      shape: 'box',
      size: 30,
    },
    physics: {
      enabled: false,
    },
  };

  const events = {
    doubleClick: (event: { nodes: number[] }): void => {
      const {
        nodes: [id],
      } = event;
      const { name } = data.nodes.find((n): boolean => n.id === id) as GraphNode;
      onSelect(name);
    },
  };

  return data.nodes.length ? (
    <GraphVis
      getNetwork={(network: { focus: (node: number, { scale }: { scale: number }) => AnyObject }): void => {
        const currentNode = data.nodes[0].id;
        network.focus(currentNode, {
          scale: 1,
        });
      }}
      graph={convertData(data)}
      events={events}
      options={options}
    />
  ) : (
    <>{`The ${type} graph of this module has no nodes.`}</>
  );
};

const Graph = ({ onSelect, componentName, type }: GraphProps): React.ReactElement => {
  const graphDataRequested = React.useRef<boolean>();
  const [state, setState] = React.useState<{ graphData: GraphData | undefined }>({ graphData: undefined });
  React.useEffect(() => {
    const graphDataListener = (event: Event, graphData: GraphData): void => {
      setState((prevState) => ({ ...prevState, graphData }));
    };
    if (!graphDataRequested.current) {
      ipcRenderer.once(`${type}-graph`, graphDataListener);
      ipcRenderer.send(`get-${type}-graph`, componentName);
      graphDataRequested.current = true;
    }

    return (): void => {
      ipcRenderer.removeAllListeners(`${type}-graph`);
    };
  });

  const { graphData } = state;

  return (
    <div className="absolute top-0 bottom-0 left-0 right-0">
      {graphData ? renderGraph(graphData, onSelect, type) : renderLoadingIcon()}
    </div>
  );
};

export default Graph;
export { GraphProps };
