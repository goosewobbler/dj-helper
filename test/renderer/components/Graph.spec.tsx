import * as React from 'react';
import { render, RenderResult, act, fireEvent } from '@testing-library/react';
import { ipcRenderer } from 'electron';
import Graph, { GraphProps } from '../../../src/renderer/components/ComponentDetails/Graph';
import { GraphData, AnyObject } from '../../../src/common/types';

type GraphOptions = {
  edges: AnyObject;
  layout: { hierarchical: AnyObject };
  nodes: AnyObject;
};

let graph: RenderResult;
let suppliedOptions: GraphOptions;
const mockGraphNetworkFocus = jest.fn();

jest.mock('../../../src/renderer/components/LoadingIcon', () => ({
  LoadingIcon: (): React.ReactElement => <p>loading</p>,
}));

jest.mock('react-graph-vis', () => ({
  __esModule: true,
  default: ({
    getNetwork,
    events: { doubleClick },
    options,
  }: {
    getNetwork: ({ focus }: { focus: typeof mockGraphNetworkFocus }) => void;
    events: { doubleClick: ({ nodes }: { nodes: number[] }) => void };
    options: GraphOptions;
  }): React.ReactElement => {
    getNetwork({
      focus: mockGraphNetworkFocus,
    });
    const clickHandler = (nodeId: number): React.MouseEventHandler => (): void => {
      doubleClick({
        nodes: [nodeId],
      });
    };
    suppliedOptions = options;
    // suppliedOptions.nodes = options.nodes;
    return (
      <>
        <div onDoubleClick={clickHandler(0)}>node 1</div>
        <div onDoubleClick={clickHandler(1)}>node 2</div>
        <div onDoubleClick={clickHandler(2)}>node 3</div>
      </>
    );
  },
}));

jest.mock('electron', () => ({
  ipcRenderer: {
    once: jest.fn(),
    send: jest.fn(),
    removeAllListeners: jest.fn(),
  },
}));

const onSelect = jest.fn();

const renderGraphWithData = (mockGraphData: GraphData, graphType: GraphProps['type']): void => {
  let sendGraphDataTimeout: ReturnType<typeof setTimeout>;
  (ipcRenderer.removeAllListeners as jest.Mock).mockImplementation(() => {
    clearTimeout(sendGraphDataTimeout);
  });
  (ipcRenderer.once as jest.Mock).mockImplementation(
    (channel, listener: (event: AnyObject, graphData: typeof mockGraphData) => void) => {
      sendGraphDataTimeout = setTimeout(() => {
        listener({}, mockGraphData);
      }, 100);
    },
  );
  graph = render(<Graph onSelect={onSelect} componentName="test-component" type={graphType} />);
};

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.useRealTimers();
});

afterEach(() => {
  graph.unmount();
});

describe('given a dependency graph', () => {
  beforeEach(() => {
    graph = render(<Graph onSelect={onSelect} componentName="test-component" type="dependency" />);
  });

  it('should make the expected data request', () => {
    expect(ipcRenderer.send).toHaveBeenCalledTimes(1);
    expect(ipcRenderer.send).toHaveBeenCalledWith('get-dependency-graph', 'test-component');
  });

  it('should render a loading icon', () => {
    expect(graph.getByText('loading')).toBeInTheDocument();
  });

  it('should render the expected html', () => {
    expect(graph.container).toMatchSnapshot();
  });
});

describe('given a dependant graph', () => {
  beforeEach(() => {
    graph = render(<Graph onSelect={onSelect} componentName="test-component" type="dependant" />);
  });

  it('should make the expected data request', () => {
    expect(ipcRenderer.send).toHaveBeenCalledTimes(1);
    expect(ipcRenderer.send).toHaveBeenCalledWith('get-dependant-graph', 'test-component');
  });

  it('should render a loading icon', () => {
    expect(graph.getByText('loading')).toBeInTheDocument();
  });

  it('should render the expected html', () => {
    expect(graph.container).toMatchSnapshot();
  });
});

describe('given a dependency graph with some data', () => {
  beforeEach(() => {
    renderGraphWithData(
      {
        edges: [
          {
            from: 0,
            to: 1,
          },
          {
            from: 0,
            to: 2,
          },
        ],
        nodes: [
          {
            id: 0,
            name: 'bbc-morph-foo',
          },
          {
            id: 1,
            name: 'bbc-morph-bar',
          },
          {
            id: 2,
            name: 'bbc-morph-baz',
          },
        ],
      },
      'dependency',
    );
  });

  it('should initially render a loading icon', () => {
    expect(graph.getByText('loading')).toBeInTheDocument();
  });

  it('should render the expected html', () => {
    expect(graph.container).toMatchSnapshot();
  });

  describe('after it receives data', () => {
    beforeEach(() => {
      act(() => {
        jest.advanceTimersByTime(101);
      });
    });

    it('should render a graph with the expected options', () => {
      expect(graph.getByText('node 1')).toBeInTheDocument();
      expect(graph.getByText('node 2')).toBeInTheDocument();
      expect(graph.getByText('node 3')).toBeInTheDocument();
      expect(suppliedOptions.layout.hierarchical).toMatchObject(expect.objectContaining({ direction: 'UD' }));
      expect(suppliedOptions).toMatchSnapshot();
    });

    it('should focus the first node in the array', () => {
      expect(mockGraphNetworkFocus).toHaveBeenCalledWith(0, { scale: 1 });
    });

    it('should call onSelect with the name of the selected node when the doubleClick handler is fired', () => {
      const node1 = graph.getByText('node 1');
      fireEvent.doubleClick(node1);
      expect(onSelect).toHaveBeenCalledWith('bbc-morph-foo');

      const node2 = graph.getByText('node 2');
      fireEvent.doubleClick(node2);
      expect(onSelect).toHaveBeenCalledWith('bbc-morph-bar');

      const node3 = graph.getByText('node 3');
      fireEvent.doubleClick(node3);
      expect(onSelect).toHaveBeenCalledWith('bbc-morph-baz');
    });

    it('should render the expected html', () => {
      expect(graph.container).toMatchSnapshot();
    });
  });
});

describe('given a dependant graph with some data', () => {
  beforeEach(() => {
    renderGraphWithData(
      {
        edges: [
          {
            from: 0,
            to: 1,
          },
          {
            from: 0,
            to: 2,
          },
        ],
        nodes: [
          {
            id: 0,
            name: 'bbc-morph-foo',
          },
          {
            id: 1,
            name: 'bbc-morph-bar',
          },
          {
            id: 2,
            name: 'bbc-morph-baz',
          },
        ],
      },
      'dependant',
    );
  });

  it('should initially render a loading icon', () => {
    expect(graph.getByText('loading')).toBeInTheDocument();
  });

  it('should render the expected html', () => {
    expect(graph.container).toMatchSnapshot();
  });

  describe('after it receives data', () => {
    beforeEach(() => {
      act(() => {
        jest.advanceTimersByTime(101);
      });
    });

    it('should render a graph with the expected options', () => {
      expect(graph.getByText('node 1')).toBeInTheDocument();
      expect(graph.getByText('node 2')).toBeInTheDocument();
      expect(graph.getByText('node 3')).toBeInTheDocument();
      expect(suppliedOptions.layout.hierarchical).toMatchObject(expect.objectContaining({ direction: 'DU' }));
      expect(suppliedOptions).toMatchSnapshot();
    });

    it('should focus the first node in the array', () => {
      expect(mockGraphNetworkFocus).toHaveBeenCalledWith(0, { scale: 1 });
    });

    it('should call onSelect with the name of the selected node when the doubleClick handler is fired', () => {
      const node1 = graph.getByText('node 1');
      fireEvent.doubleClick(node1);
      expect(onSelect).toHaveBeenCalledWith('bbc-morph-foo');

      const node2 = graph.getByText('node 2');
      fireEvent.doubleClick(node2);
      expect(onSelect).toHaveBeenCalledWith('bbc-morph-bar');

      const node3 = graph.getByText('node 3');
      fireEvent.doubleClick(node3);
      expect(onSelect).toHaveBeenCalledWith('bbc-morph-baz');
    });

    it('should render the expected html', () => {
      expect(graph.container).toMatchSnapshot();
    });
  });
});

describe('given a dependant graph with some data but no nodes', () => {
  beforeEach(() => {
    renderGraphWithData(
      {
        edges: [],
        nodes: [],
      },
      'dependant',
    );
  });

  it('should initially render a loading icon', () => {
    expect(graph.getByText('loading')).toBeInTheDocument();
  });

  it('should render the expected html', () => {
    expect(graph.container).toMatchSnapshot();
  });

  describe('after it receives data', () => {
    beforeEach(() => {
      act(() => {
        jest.advanceTimersByTime(101);
      });
    });

    it('should render the expected messsge', () => {
      expect(graph.container).toHaveTextContent('The dependant graph of this module has no nodes.');
    });

    it('should render the expected html', () => {
      expect(graph.container).toMatchSnapshot();
    });
  });
});

describe('given a dependency graph with some data but no nodes', () => {
  beforeEach(() => {
    renderGraphWithData(
      {
        edges: [],
        nodes: [],
      },
      'dependency',
    );
  });

  it('should initially render a loading icon', () => {
    expect(graph.getByText('loading')).toBeInTheDocument();
  });

  it('should render the expected html', () => {
    expect(graph.container).toMatchSnapshot();
  });

  describe('after it receives data', () => {
    beforeEach(() => {
      act(() => {
        jest.advanceTimersByTime(101);
      });
    });

    it('should render the expected messsge', () => {
      expect(graph.container).toHaveTextContent('The dependency graph of this module has no nodes.');
    });

    it('should render the expected html', () => {
      expect(graph.container).toMatchSnapshot();
    });
  });
});
