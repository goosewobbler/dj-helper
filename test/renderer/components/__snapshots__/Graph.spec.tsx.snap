// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given a dependant graph should render the expected html 1`] = `
<div>
  <div
    class="absolute top-0 bottom-0 left-0 right-0"
  >
    <div
      class="flex items-center justify-center w-full h-full loading"
    >
      <div
        class="w-8 h-8"
      >
        <p>
          loading
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`given a dependant graph with some data after it receives data should render a graph with the expected options 1`] = `
Object {
  "edges": Object {
    "arrows": Object {
      "from": Object {
        "enabled": true,
        "scaleFactor": 0.5,
      },
      "to": Object {
        "enabled": false,
        "scaleFactor": 0.5,
      },
    },
    "smooth": Object {
      "forceDirection": "vertical",
      "roundness": 1,
      "type": "cubicBezier",
    },
  },
  "layout": Object {
    "hierarchical": Object {
      "direction": "DU",
      "edgeMinimization": true,
      "nodeSpacing": 350,
      "parentCentralization": true,
      "sortMethod": "directed",
    },
    "improvedLayout": true,
  },
  "nodes": Object {
    "color": Object {},
    "font": Object {},
    "scaling": Object {
      "max": 30,
      "min": 30,
    },
    "shape": "box",
    "size": 30,
  },
  "physics": Object {
    "enabled": false,
  },
}
`;

exports[`given a dependant graph with some data after it receives data should render the expected html 1`] = `
<div>
  <div
    class="absolute top-0 bottom-0 left-0 right-0"
  >
    <div>
      node 1
    </div>
    <div>
      node 2
    </div>
    <div>
      node 3
    </div>
  </div>
</div>
`;

exports[`given a dependant graph with some data but no nodes after it receives data should render the expected html 1`] = `
<div>
  <div
    class="absolute top-0 bottom-0 left-0 right-0"
  >
    The dependant graph of this module has no nodes.
  </div>
</div>
`;

exports[`given a dependant graph with some data but no nodes should render the expected html 1`] = `
<div>
  <div
    class="absolute top-0 bottom-0 left-0 right-0"
  >
    <div
      class="flex items-center justify-center w-full h-full loading"
    >
      <div
        class="w-8 h-8"
      >
        <p>
          loading
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`given a dependant graph with some data should render the expected html 1`] = `
<div>
  <div
    class="absolute top-0 bottom-0 left-0 right-0"
  >
    <div
      class="flex items-center justify-center w-full h-full loading"
    >
      <div
        class="w-8 h-8"
      >
        <p>
          loading
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`given a dependency graph should render the expected html 1`] = `
<div>
  <div
    class="absolute top-0 bottom-0 left-0 right-0"
  >
    <div
      class="flex items-center justify-center w-full h-full loading"
    >
      <div
        class="w-8 h-8"
      >
        <p>
          loading
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`given a dependency graph with some data after it receives data should render a graph with the expected options 1`] = `
Object {
  "edges": Object {
    "arrows": Object {
      "from": Object {
        "enabled": false,
        "scaleFactor": 0.5,
      },
      "to": Object {
        "enabled": true,
        "scaleFactor": 0.5,
      },
    },
    "smooth": Object {
      "forceDirection": "vertical",
      "roundness": 1,
      "type": "cubicBezier",
    },
  },
  "layout": Object {
    "hierarchical": Object {
      "direction": "UD",
      "edgeMinimization": true,
      "nodeSpacing": 350,
      "parentCentralization": true,
      "sortMethod": "directed",
    },
    "improvedLayout": true,
  },
  "nodes": Object {
    "color": Object {},
    "font": Object {},
    "scaling": Object {
      "max": 30,
      "min": 30,
    },
    "shape": "box",
    "size": 30,
  },
  "physics": Object {
    "enabled": false,
  },
}
`;

exports[`given a dependency graph with some data after it receives data should render the expected html 1`] = `
<div>
  <div
    class="absolute top-0 bottom-0 left-0 right-0"
  >
    <div>
      node 1
    </div>
    <div>
      node 2
    </div>
    <div>
      node 3
    </div>
  </div>
</div>
`;

exports[`given a dependency graph with some data but no nodes after it receives data should render the expected html 1`] = `
<div>
  <div
    class="absolute top-0 bottom-0 left-0 right-0"
  >
    The dependency graph of this module has no nodes.
  </div>
</div>
`;

exports[`given a dependency graph with some data but no nodes should render the expected html 1`] = `
<div>
  <div
    class="absolute top-0 bottom-0 left-0 right-0"
  >
    <div
      class="flex items-center justify-center w-full h-full loading"
    >
      <div
        class="w-8 h-8"
      >
        <p>
          loading
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`given a dependency graph with some data should render the expected html 1`] = `
<div>
  <div
    class="absolute top-0 bottom-0 left-0 right-0"
  >
    <div
      class="flex items-center justify-center w-full h-full loading"
    >
      <div
        class="w-8 h-8"
      >
        <p>
          loading
        </p>
      </div>
    </div>
  </div>
</div>
`;
